enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+
+#### Changes to NTO
+
+The NTO PerformanceProfile will be updated to support a new flag which will
+toggle the workload pinning to the `reserved` cores. The idea here being to
+simplify the approach for how customers set this configuration. With PAO being
+part of NTO now ([see here for more info](../node-tuning/pao-in-nto.md)) this
+affords us the chance to consolidate the configuration for `kubelet` and `crio`.
+
+We will modify the code path that generates the [new machine
+config](https://github.com/openshift/cluster-node-tuning-operator/blob/a780dfe07962ad07e4d50c852047ef8cf7b287da/pkg/performanceprofile/controller/performanceprofile/components/machineconfig/machineconfig.go#L91-L127)
+using the performance profile. With the new `spec.workloads[Infrastructure]`
+enum we will add the configuration for `crio` and `kubelet` to the final machine
+config manifest. Then the existing code path will apply the change as normal.
+
+#### API Server Admission Hook
+
+We will need to alter the code in the admission controller to remove the check
+for Single Node Topology, and modify the check for running nodes to check the
+global identifier which will be set at install time.
+
+The existing admission hook has 4 checks when it comes to workload pinning.
+
+Old Path:
+
+1. Check if `pod` is a static pod
+   - Skips modification attempt if it is static.
+2. Checks if currently running cluster topology is Single Node - **Will Change**
+   - Skips modification if it is anything other than Single Node
+3. Checks if all running nodes are managed - **Will Change**
+   - Skips modification if any of the nodes are not managed
+4. Checks what resource limits and requests are set on the pod
+   - Skips modification if QoS is guaranteed or both limits and requests are set
+   - Skips modification if after update the QoS is changed
+
+Changed Path:
+
+1. Check if `pod` is a static pod
+   - Skips modification attempt if it is static.
+2. Checks if currently running cluster has global identifier for partitioning
+   set
+   - Skips modification if identifier partitioning set to `None` unless Single
+     Node, will check with old logic to maintain upgrade for Single-Node 4.11 ->
+     4.12.
+3. Checks what resource limits and requests are set on the pod
+   - Skips modification if QoS is guaranteed or both limits and requests are set
+   - Skips modification if after update the QoS is changed
+
+### Risks and Mitigations
+
+The sames risks and mitigations highlighted in [Management Workload
+Partitioning](management-workload-partitioning.md) apply to this enhancement as
+well.
+
+We need to make it very clear to customers that this feature is supported as a
+day 0 configuration and day n+1 alterations are not be supported with this
+enhancement. Part of that messaging should involve a clear indication that this
+currently will be a cluster wide feature.
+
+A risk we can run into is that a customer can apply a CPU set that is too small
```

Comment:
How will the user undo a mistake? Mistake in this are can cause the node to not boot or join the cluster again. And there must be a way out of this. Especially on SNO and 3NC where you can end up without the api server that holds the manifests that must be fixed.

Reply:
In regard to this one, after some testing there are some headaches with the typical recovery techniques for 3NC and normal clusters. I think the best approach is to stop performance profile updates that are out of bounds on the reserved CPUSet by checking the CPU capacity of the nodes for the MCP. On larger clusters, for performance reasons, we only check at most 3 nodes per pool, since we require that all MCPs are homogeneous in CPU size. We can add that in this EP, but I think we can do it as it's own task for NTO, what do you think?

Reply:
I think we need to ask @jmencak if NTO level validation is what you are proposing. There were NTO related bugs like that too in the past.

However in general, validators are tricky as you need to think about a case where joins the cluster later, when the tuned already exists. But iirc that was being handled in different part of this enhancement, right?

Reply:
> I think we need to ask @jmencak if NTO level validation is what you are proposing. There were NTO related bugs like that too in the past.

This seems to be a frequently requested feature, so it makes sense at least look into this.  Adding to our backlog.

Reply:
Just for reference. This was one of the bugs where a heterogenous cluster got into a reboot loop: https://issues.redhat.com/browse/OCPBUGS-646

Reply:
I see, so @MarSik is the current documentation on heterogenous MCPs good enough for you to be okay with this Epic? 

I'm thinking having a separate ticket to add validation support would make it easier to backport, if need be, thoughts?

Reply:
+1